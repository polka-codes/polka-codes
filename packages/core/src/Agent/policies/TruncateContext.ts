// TruncateContext policy - generated by polka.codes
// Truncates conversation messages when context size becomes too large to prevent token limit overflows

import type { MessageParam } from '../../AiService'
import { Policies } from '../../config'
import type { AgentBase, AgentPolicy } from '../AgentBase'

// Default max token threshold
const DEFAULT_MAX_TOKENS_ESTIMATE = 4000

// Simple token estimation (1 token â‰ˆ 4 characters)
function estimateTokens(text: string): number {
  return Math.ceil(text.length / 4)
}

export const TruncateContextPolicy = (tools: Parameters<AgentPolicy>[0]) => {
  return {
    name: Policies.TruncateContext,
    async onBeforeRequest(agent: AgentBase): Promise<void> {
      const messages = agent.messages

      // Skip if there aren't enough messages to truncate
      if (messages.length < 3) {
        return
      }

      // Calculate total token estimate
      let totalTokens = 0
      for (const msg of messages) {
        if (typeof msg.content === 'string') {
          totalTokens += estimateTokens(msg.content)
        } else if (Array.isArray(msg.content)) {
          // Handle content blocks
          for (const block of msg.content) {
            if (typeof block === 'object' && 'text' in block && typeof block.text === 'string') {
              totalTokens += estimateTokens(block.text)
            }
          }
        }
      }

      // If under threshold, no truncation needed
      if (totalTokens <= DEFAULT_MAX_TOKENS_ESTIMATE) {
        return
      }

      // Truncation strategy: preserve recent messages
      // Keep the most recent messages (last 6 messages = 3 exchanges)
      const keepLastN = 6
      const recentMessages = messages.slice(-keepLastN)

      // Count how many messages were truncated
      const truncatedCount = messages.length - recentMessages.length

      // Rebuild the truncated message list
      const truncatedMessages: MessageParam[] = [
        // Add a message explaining truncation if any messages were removed
        ...(truncatedCount > 0
          ? [
              {
                role: 'user' as const,
                content: `Note: ${truncatedCount} earlier messages were truncated to prevent context overflow.`,
              },
            ]
          : []),
        ...recentMessages,
      ]

      // Update the agent's messages
      agent.setMessages(truncatedMessages)
    },
  }
}
