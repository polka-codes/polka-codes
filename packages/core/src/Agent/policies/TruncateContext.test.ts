// TruncateContext policy tests - generated by polka.codes

import { expect, spyOn, test } from 'bun:test'
import type { MessageParam } from '../../AiService'
import { TruncateContextPolicy } from './TruncateContext'

test('TruncateContextPolicy should truncate messages when over threshold', async () => {
  // Create mock agent
  const mockAgent = {
    messages: [
      { role: 'user', content: 'User message 1'.repeat(500) },
      { role: 'assistant', content: 'Assistant message 1'.repeat(500) },
      { role: 'user', content: 'User message 2' },
      { role: 'assistant', content: 'Assistant message 2' },
      { role: 'user', content: 'User message 3' },
      { role: 'assistant', content: 'Assistant message 3' },
    ] as MessageParam[],
    setMessages: (messages: MessageParam[]) => {
      mockAgent.messages = messages
    },
  }

  const policy = TruncateContextPolicy({})
  await policy.onBeforeRequest(mockAgent as any)

  expect(mockAgent.messages).toMatchSnapshot()
})

test('TruncateContextPolicy should not truncate small message lists', async () => {
  const mockAgent = {
    messages: [
      { role: 'user', content: 'Short user message' },
      { role: 'assistant', content: 'Short assistant message' },
    ] as MessageParam[],
    setMessages: (messages: MessageParam[]) => {
      mockAgent.messages = messages
    },
  }

  const setMessagesSpy = spyOn(mockAgent, 'setMessages')
  const policy = TruncateContextPolicy({})
  await policy.onBeforeRequest(mockAgent as any)

  expect(setMessagesSpy).not.toHaveBeenCalled()
  expect(mockAgent.messages.length).toBe(2)
})

test('TruncateContextPolicy should preserve recent exchanges', async () => {
  const mockAgent = {
    messages: [
      { role: 'user', content: 'Old user message 1'.repeat(300) },
      { role: 'assistant', content: 'Old assistant message 1'.repeat(300) },
      { role: 'user', content: 'Old user message 2'.repeat(300) },
      { role: 'assistant', content: 'Old assistant message 2'.repeat(300) },
      { role: 'user', content: 'Recent user message 1' },
      { role: 'assistant', content: 'Recent assistant message 1' },
      { role: 'user', content: 'Recent user message 2' },
      { role: 'assistant', content: 'Recent assistant message 2' },
      { role: 'user', content: 'Most recent user message' },
      { role: 'assistant', content: 'Most recent assistant message' },
    ] as MessageParam[],
    setMessages: (messages: MessageParam[]) => {
      mockAgent.messages = messages
    },
  }

  const policy = TruncateContextPolicy({})
  await policy.onBeforeRequest(mockAgent as any)

  expect(mockAgent.messages).toMatchSnapshot()
})
