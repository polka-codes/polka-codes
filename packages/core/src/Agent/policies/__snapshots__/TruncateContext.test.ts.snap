// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`TruncateContextPolicy should truncate messages when over threshold 1`] = `
[
  {
    "content": "User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1User message 1",
    "role": "user",
  },
  {
    "content": "Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1Assistant message 1",
    "role": "assistant",
  },
  {
    "content": "User message 2",
    "role": "user",
  },
  {
    "content": "Assistant message 2",
    "role": "assistant",
  },
  {
    "content": "User message 3",
    "role": "user",
  },
  {
    "content": "Assistant message 3",
    "role": "assistant",
  },
]
`;

exports[`TruncateContextPolicy should preserve recent exchanges 1`] = `
[
  {
    "content": "Note: 4 earlier messages were truncated to prevent context overflow.",
    "role": "user",
  },
  {
    "content": "Recent user message 1",
    "role": "user",
  },
  {
    "content": "Recent assistant message 1",
    "role": "assistant",
  },
  {
    "content": "Recent user message 2",
    "role": "user",
  },
  {
    "content": "Recent assistant message 2",
    "role": "assistant",
  },
  {
    "content": "Most recent user message",
    "role": "user",
  },
  {
    "content": "Most recent assistant message",
    "role": "assistant",
  },
]
`;
